### PR TITLE
Add GET /api/core/v3/validators routes

### LINKED ISSUE #1053: Add `/api/core/v3/validators` route
https://github.com/iotaledger/tips-draft/blob/43b25b17b564bc42963d9f32645a98f1bc2a58c3/tips/TIPS-API/openapi3-core.yaml#L259C3-L259C26

### LINKED ISSUE #1054: Add `/api/core/v3/validators/{accountId}` route
https://github.com/iotaledger/tips-draft/blob/43b25b17b564bc42963d9f32645a98f1bc2a58c3/tips/TIPS-API/openapi3-core.yaml#L299C3-L299C38

### PR DESCRIPTION
**Pull request by kwek20**
Fixes #1053 and #1054

### CODE REVIEW

File: sdk/src/client/node_api/core/routes.rs [hunk 1]
```diff
@@ -103,6 +106,30 @@ impl ClientInner {
             .await
     }
 
+    /// Returns information of all registered validators and if they are active.
+    /// GET JSON to /api/core/v3/validators
+    pub async fn get_validators(&self) -> Result<ValidatorsResponse> {
```
Comment (thibault-martinez) [inline sdk/src/client/node_api/core/routes.rs]: Aren't you the one that suggested we respect the API order :trollface: ?

File: sdk/src/client/node_api/core/routes.rs [hunk 2]
```diff
@@ -15,6 +15,9 @@ use crate::{
     },
     types::{
         api::core::response::{
+            AccountStakingResponse, BlockMetadataResponse, CongestionResponse, InfoResponse,
+            IssuanceBlockHeaderResponse, PeerResponse, RoutesResponse, SubmitBlockResponse, UtxoChangesResponse,
+            ValidatorsResponse,
             BlockMetadataResponse, CongestionResponse, InfoResponse, IssuanceBlockHeaderResponse, ManaRewardsResponse,
```
Comment (Thoralf-M) [inline sdk/src/client/node_api/core/routes.rs]: Duplicated imports

**Comment by thibault-martinez** [comment]
Can be undrafted ?

File: sdk/src/client/node_api/core/routes.rs [hunk 3]
```diff
@@ -122,6 +123,31 @@ impl ClientInner {
             .await
     }
 
+    /// Returns information of all registered validators and if they are active.
+    /// GET JSON to /api/core/v3/validators
+    pub async fn get_validators(&self) -> Result<ValidatorsResponse> {
+        // TODO: Pagination is not in the TIP yet
```
Comment (Thoralf-M) [inline sdk/src/client/node_api/core/routes.rs]: Will there be pagination?

**Review by thibault-martinez — CHANGES_REQUESTED** [review]
Conflicts

File: sdk/src/client/node_api/core/routes.rs [hunk 4]
```diff
@@ -129,14 +130,38 @@ impl ClientInner {
...
         self.node_manager
             .read()
             .await
             .get_request(PATH, epoch_index.as_deref(), self.get_timeout().await, false, false)
             .await
     }
 
+    /// Returns information of all registered validators and if they are active.
+    /// GET JSON to /api/core/v3/validators
+    pub async fn get_validators(&self, page_size: Option<u32>) -> Result<ValidatorsResponse> {
```
Comment (Thoralf-M) [inline sdk/src/client/node_api/core/routes.rs:146]: page_size alone is useless, we would also need to be able to provide a cursor https://github.com/iotaledger/tips-draft/pull/24/files#r1308764062

Comment (kwek20) [reply sdk/src/client/node_api/core/routes.rs:146]: Why return the page_size and cursor as well then? seems weird to me :/

Comment (Thoralf-M) [reply sdk/src/client/node_api/core/routes.rs:146]: It needs to be returned so you know what you have to provide for the next page (or pagination just gets removed altogether for this endpoint)

**Review by thibault-martinez — CHANGES_REQUESTED** [review]
Conflicts

File: sdk/src/client/node_api/core/routes.rs [hunk 5]
```diff
@@ -129,14 +130,38 @@ impl ClientInner {
...
+        self.node_manager
+            .read()
+            .await
+            .get_request(PATH, page_size.as_deref(), self.get_timeout().await, false, false)
+            .await
+    }
+
+    /// Return the information of requested staker.
+    /// GET /api/core/v3/validators/{accountId}
+    pub async fn get_validator(&self, account_id: &AccountId) -> Result<AccountStakingResponse> {
```
Comment (Alex6323) [inline sdk/src/client/node_api/core/routes.rs]: Something doesn't feel right to me here (I will also mention this in the TIP draft):
*  The docs and the return type seem to suggest a name like: `get_staker` or `get_staker_info`
*  But the endpoint route seems to suggest that `get_validator` or `get_validator_info` is the more appropriate name, but then the docs probably need to change.

File: sdk/src/client/node_api/core/routes.rs (hunk 4 above)
Comment (Thoralf-M) [reply sdk/src/client/node_api/core/routes.rs:146]: https://github.com/iotaledger/iota-sdk/issues/1118

**Review by Thoralf-M — DISMISSED** [review]

File: sdk/src/client/node_api/core/routes.rs (hunk 5 above)
Comment (thibault-martinez) [reply sdk/src/client/node_api/core/routes.rs]: Ask things like that in the TIP, if this PR is a strict implementation of the current state of the TIP, we'll move forward with it

File: sdk/src/types/api/core/response.rs [hunk 6]
```diff
@@ -165,10 +165,38 @@ pub struct BaseTokenResponse {
     pub use_metric_prefix: bool,
 }
 
+/// Response of
+/// - GET /api/core/v3/blocks/validators
+/// A paginated list of all registered validators ready for the next epoch and indicates if they were active recently
+/// (are eligible for committee selection).
+#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
+pub struct ValidatorsResponse {
```
Comment (DaughterOfMars) [inline sdk/src/types/api/core/response.rs:176]: Needs camelCase here and below?

File: sdk/src/types/api/core/response.rs [hunk 7]
```diff
@@ -165,10 +165,38 @@ pub struct BaseTokenResponse {
...
+    /// last page.
+    cursor: String,
+}
+
+/// Response of GET /api/core/v3/rewards/{outputId}.
+/// Returns the mana rewards of an account or delegation output.
+#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
+pub struct ManaRewardsResponse {
+    /// The starting epoch index for which the mana rewards are returned.
+    pub epoch_start: u64, // TODO: replace with `EpochIndex`
```
Comment (DaughterOfMars) [inline sdk/src/types/api/core/response.rs]: These can be replaced

File: sdk/src/types/api/core/response.rs [hunk 8]
```diff
@@ -199,6 +226,33 @@ pub struct CommitteeMember {
...
 
+/// Information of a validator.
+#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
+pub struct Validator {
+    /// The account identifier of the validator
+    account_id: AccountId,
+    /// The epoch index until which the validator registered to stake.
+    staking_end_epoch: EpochIndex,
+    /// The total stake of the pool, including delegators.
+    #[serde(with = "crate::utils::serde::string")]
```
Comment (DaughterOfMars) [inline sdk/src/types/api/core/response.rs]: Can import these paths

File: sdk/src/types/api/core/response.rs [hunk 9]
```diff
@@ -165,40 +168,96 @@ pub struct BaseTokenResponse {
...
     pub validator_stake: u64,
     /// The fixed cost of the validator, which it receives as part of its Mana rewards.
-    #[serde(with = "crate::utils::serde::string")]
+    #[serde(with = "string")]
     pub fixed_cost: u64,
 }
 
+/// Information of a validator.
+#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
+pub struct Validator {
```
Comment (Thoralf-M) [inline sdk/src/types/api/core/response.rs]: Also camelCase here?

**Review by DaughterOfMars — DISMISSED** [review]

File: sdk/src/client/node_api/core/routes.rs (hunk 5 above)
Comment (thibault-martinez) [reply sdk/src/client/node_api/core/routes.rs]: This has been changed

**Review by Thoralf-M — APPROVED** [review]

**Review by Alex6323 — APPROVED** [review]

**Review by thibault-martinez — APPROVED** [review]

**Review by abdulmth — APPROVED** [review]

**Pull request merged**